tdxctl/src/fde_setup.rs

```diff
@@ -181,44 +184,111 @@ impl SetupFdeArgs {
-            cmd_gen_ra_cert(GenRaCertArgs {
-                ca_cert: host_shared.dir.tmp_ca_cert_file(),
-                ca_key: host_shared.dir.tmp_ca_key_file(),
-                cert_path: gen_certs_dir.join("cert.pem"),
-                key_path: gen_certs_dir.join("key.pem"),
-            })?;
-            info!("Requesting app keys from KMS: {kms_url}");
-            let ra_client = RaClient::new_mtls(
-                format!("{kms_url}/prpc"),
-                fs::read_to_string(host_shared.dir.kms_ca_cert_file())?,
-                fs::read_to_string(gen_certs_dir.join("cert.pem"))?,
-                fs::read_to_string(gen_certs_dir.join("key.pem"))?,
-            )?;
-            let kms_client = kms_rpc::kms_client::KmsClient::new(ra_client);
-            let response = kms_client
-                .get_app_key(GetAppKeyRequest { upgradable: true })
+    async fn request_app_keys_from_kms(&self, host_shared: &HostShared) -> Result<()> {
+        let Some(kms_url) = &host_shared.vm_config.kms_url else {
+            bail!("KMS URL is not set");
+        };
+        info!("Requesting app keys from KMS: {kms_url}");
+        let gen_certs_dir = self.work_dir.join("certs");
+        fs::create_dir_all(&gen_certs_dir).context("Failed to create certs dir")?;
+        let kms_url = format!("{kms_url}/prpc");
+
+        let tmp_ca = {
+            info!("Getting temp ca cert");
+            let client = RaClient::new(kms_url.clone(), true)?;
+            let kms_client = kms_rpc::kms_client::KmsClient::new(client);
+            kms_client
+                .get_temp_ca_cert()
                 .await
-                .context("Failed to get app key")?;
-            let keys_json =
-                serde_json::to_string(&response).context("Failed to serialize app keys")?;
-            fs::write(self.app_keys_file(), keys_json).context("Failed to write app keys")?;
-        } else {
-            info!("KMS is not enabled, generating local app keys");
-            cmd_gen_app_keys(GenAppKeysArgs {
-                ca_level: 1,
-                output: self.app_keys_file(),
-            })?;
+                .context("Failed to get temp ca cert")?
+        };
+        let cert_pair = generate_ra_cert(tmp_ca.temp_ca_cert, tmp_ca.temp_ca_key)?;
+        let ra_client = RaClientConfig::builder()
+            .tls_no_check(false)
+            .tls_built_in_root_certs(false)
+            .remote_uri(kms_url.clone())
+            .tls_client_cert(cert_pair.cert_pem)
+            .tls_client_key(cert_pair.key_pem)
+            .tls_ca_cert(tmp_ca.ca_cert.clone())
+            .cert_validator(Box::new(|cert| {
+                let Some(cert) = cert else {
+                    bail!("Missing server cert");
+                };
+                let Some(usage) = cert.special_usage else {
+                    bail!("Missing server cert usage");
+                };
+                if usage != "kms:rpc" {
```

Comment:
Review Hint: A real kms only issues certs with `usage=kms:rpc` to themselves to server rpc, nerver issues it to apps.